src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Helpers.cs

```diff
@@ -36,37 +35,57 @@ public static partial class JsonSerializer
             return success;
         }
 
-        private static void WriteUsingMetadata<TValue>(Utf8JsonWriter writer, in TValue value, JsonTypeInfo jsonTypeInfo)
+        private static void WriteUsingGeneratedSerializer<TValue>(Utf8JsonWriter writer, in TValue value, JsonTypeInfo jsonTypeInfo)
         {
-            if (jsonTypeInfo is JsonTypeInfo<TValue> typedInfo &&
-                typedInfo.Serialize != null &&
+            if (jsonTypeInfo.HasSerialize &&
+                jsonTypeInfo is JsonTypeInfo<TValue> typedInfo &&
                 typedInfo.Options._context?.CanUseSerializationLogic == true)
             {
+                Debug.Assert(typedInfo.Serialize != null);
                 typedInfo.Serialize(writer, value);
                 writer.Flush();
             }
             else
             {
-                WriteStack state = default;
-                state.Initialize(jsonTypeInfo, supportContinuation: false);
+                WriteUsingSerializer(writer, value, jsonTypeInfo);
+            }
+        }
+
+        private static void WriteUsingSerializer<TValue>(Utf8JsonWriter writer, in TValue value, JsonTypeInfo jsonTypeInfo)
+        {
+            Debug.Assert(jsonTypeInfo is not JsonTypeInfo<TValue>, "Incorrect method called.");
```

Comment:
```suggestion
            Debug.Assert(writer is not null);
            Debug.Assert(jsonTypeInfo is not JsonTypeInfo<TValue>, "Incorrect method called.");
```

Reply:
Because writer can be passed in by the user?